File: p22.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<string> ret;
    void go(string str, int s, int e, int depth){
        if(e == depth){
            ret.push_back(str);
            return;
        }
        if(s > depth) return;
        if(s > e){
            go(str + ")", s, e + 1, depth);
        }
        go(str + "(", s + 1, e, depth);
    }
    vector<string> generateParenthesis(int n) {
        go("", 0, 0, n);
        return ret;
    }
};
int main(){
    int N;
    Solution s;
    cin >> N;
    vector<string> ret = s.generateParenthesis(N);
    for(string s : ret) cout << s << "\n";
}

File: p22.java
import java.util.ArrayList;
import java.util.List;

public class p22 {
    public static List<String> ret;
    public List<String> generateParenthesis(int n) {
        ret = new ArrayList<>();
        
        go("", 0, 0, n);

        return ret;
    }

    public static void go(String str, int l, int r, int size){
        if(l == size && r == size){
            ret.add(str);
            return;
        }

        if(l < size){
            go(str + "(", l + 1, r, size);
        }
        if(r < l){
            go(str + ")", l, r + 1, size);
        }
    }
}
